### PR TITLE
Replay history from file

### PR DESCRIPTION
**Pull request by halakaraki**
Issue #274 Add a method to allow replaying history from a json file (dowloaded from cli or ui)

### CODE REVIEW

File: worker/worker.go [hunk 1]
```diff
@@ -91,6 +91,17 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
 	return internal.ReplayWorkflowHistory(logger, history)
 }
 
+// ReplayWorkflowHistoryFromJSONFile executes a single decision task for the given history.
+// Use for testing the backwards compatibility of code changes and troubleshooting workflows in a debugger.
+//
+// The logger is an optional parameter. Defaults to the noop logger.
+// The response contains the decisions produced processing the decision task. It is either
```
Comment (mfateev) [inline worker/worker.go]: There is no response besides error in this function signature.

File: internal/worker.go [hunk 2]
```diff
@@ -304,3 +292,24 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
...
+	raw, err := ioutil.ReadFile(jsonfileName)
+	if err != nil {
+		return nil, err
+	}
+
+	var deserializedEvents []shared.HistoryEvent
+	json.Unmarshal(raw, &deserializedEvents)
+
+	eventCount := len(deserializedEvents)
+	historyEvents := make([]*shared.HistoryEvent, eventCount, eventCount)
```
Comment (mfateev) [inline internal/worker.go]: Why is copy of history events needed here?

File: internal/worker.go [hunk 3]
```diff
@@ -200,57 +202,45 @@ func ReplayWorkflowExecution(ctx context.Context, service workflowserviceclient.
...
+// The logger is an optional parameter. Defaults to the noop logger.
+func ReplayWorkflowHistoryFromJSONFile(logger *zap.Logger, jsonfileName string) error {
+
+	history, err := extractHistoryFromFile(jsonfileName)
+
+	if err != nil {
+		return err
+	}
+
+	testReporter := logger.Sugar()
```
Comment (mfateev) [inline internal/worker.go]: Comment says that logger is optional. I believe it is going to panic if logger is nil here.

File: internal/testdata/sampleHistory.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
+[
```
Comment (mfateev) [inline internal/testdata/sampleHistory.json]: Is this the format produced by UI?

Comment (halakaraki) [reply internal/testdata/sampleHistory.json]: It's the one produced by downloading the file from the cli

File: worker/worker.go (hunk 1 above)
Comment (halakaraki) [reply worker/worker.go]: fixed the comment

File: internal/worker.go (hunk 3 above)
Comment (halakaraki) [reply internal/worker.go]: Added null check

File: internal/worker.go (hunk 2 above)
Comment (halakaraki) [reply internal/worker.go]: Resolved

File: internal/testdata/sampleHistory.json (hunk 4 above)
Comment (halakaraki) [reply internal/testdata/sampleHistory.json]: The UI format is different than what's produced by the api. This is tracked by https://github.com/uber/cadence-web/issues/19

File: internal/testdata/sampleHistory.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    "eventId": 5,
+    "eventType": "ActivityTaskScheduled",
+    "activityTaskScheduledEventAttributes": {
+      "activityId": "0",
+      "activityType": {
+        "name": "testActivity"
+      },
+      "taskList": {
+        "name": "taskList1"
+      }
```
Comment (nathanboktae) [inline internal/testdata/sampleHistory.json]: Did you test to see if `input` and other binary fields are properly being round tripped?

File: internal/worker.go [hunk 6]
```diff
@@ -200,57 +202,53 @@ func ReplayWorkflowExecution(ctx context.Context, service workflowserviceclient.
...
+	controller := gomock.NewController(testReporter)
+	service := workflowservicetest.NewMockClient(controller)
+
 	domain := "ReplayDomain"
+
+	return replayWorkflowHistory(logger, service, domain, history)
+}
+
+func replayWorkflowHistory(logger *zap.Logger, service workflowserviceclient.Interface, domain string, history *shared.History) error {
```
[the remark this answers is not in the document]
Comment (arthurguber) [reply internal/worker.go]: is "logger" argument used anywhere in this new refactored "replayWorkflowHistory"?

File: internal/testdata/sampleHistory.json (hunk 5 above)
Comment (halakaraki) [reply internal/testdata/sampleHistory.json]: changed it to a sample history that has both input and timeouts.

File: internal/worker.go [hunk 7]
```diff
@@ -304,3 +300,18 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
...
+
+func extractHistoryFromFile(jsonfileName string) (*shared.History, error) {
+
+	raw, err := ioutil.ReadFile(jsonfileName)
+	if err != nil {
+		return nil, err
+	}
+
+	var deserializedEvents []*shared.HistoryEvent
+	json.Unmarshal(raw, &deserializedEvents)
```
Comment (mfateev) [inline internal/worker.go]: The error returned from Unmarshal is ignored

File: worker/worker.go [hunk 8]
```diff
@@ -80,17 +80,22 @@ func EnableVerboseLogging(enable bool) {
...
 
-// ReplayWorkflowHistory executes a single decision task for the given history.
-// Use for testing the backwards compatibility of code changes and troubleshooting workflows in a debugger.
-//
-// The logger is an optional parameter. Defaults to the noop logger.
-// The response contains the decisions produced processing the decision task. It is either
-// RespondDecisionTaskCompletedRequest or RespondDecisionTaskFailedRequest.
-// The returned stackTrace contains the stack trace of the workflow at the end of the decision.
+// ReplayWorkflowHistory executes a single decision task for the given json history file.
+////// Use for testing the backwards compatibility of code changes and troubleshooting workflows in a debugger.
```
Comment (mfateev) [inline worker/worker.go]: What are these /////?

File: internal/worker.go [hunk 9]
```diff
@@ -304,3 +300,21 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
...
+func extractHistoryFromFile(jsonfileName string) (*shared.History, error) {
+
+	raw, err := ioutil.ReadFile(jsonfileName)
+	if err != nil {
+		return nil, err
+	}
+
+	var deserializedEvents []*shared.HistoryEvent
+	err = json.Unmarshal(raw, &deserializedEvents)
+
```
Comment (mfateev) [inline internal/worker.go]: remove empty line

File: internal/worker.go [hunk 10]
```diff
@@ -304,3 +300,21 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
 	_, _, err := taskHandler.ProcessWorkflowTask(task, iterator)
 	return err
 }
+
+func extractHistoryFromFile(jsonfileName string) (*shared.History, error) {
+
```
Comment (mfateev) [inline internal/worker.go]: remove empty line

File: internal/worker.go [hunk 11]
```diff
@@ -304,3 +300,21 @@ func ReplayWorkflowHistory(logger *zap.Logger, history *shared.History) error {
...
+	var deserializedEvents []*shared.HistoryEvent
+	err = json.Unmarshal(raw, &deserializedEvents)
+
+	if err != nil {
+		return nil, err
+	}
+	history := &shared.History{Events: deserializedEvents}
+
+	return history, nil
+}
```
Comment (mfateev) [inline internal/worker.go]: add new line at the end of the file

**Pull request merged**